cost/azure/databricks/rightsize_compute/azure_databricks_rightsize_compute.pt

```diff
@@ -0,0 +1,1478 @@
+
+    // If we're checking both, do an 'and' or an 'or' depending on the value of param_stats_check_both
+    if (checking_cpu && checking_mem) {
+      if (param_stats_check_both == "Both CPU and Memory") {
+        is_idle = is_idle_cpu && is_idle_mem
+        is_underutil = is_underutil_cpu && is_underutil_mem
+      } else {
+        is_idle = is_idle_cpu || is_idle_mem
+        is_underutil = is_underutil_cpu || is_underutil_mem
+      }
+    }
+
+    // Only bother doing anything if we're checking at least one metric
+    if (checking_cpu || checking_mem) {
+      if (is_idle) { // First check to see if the resource is below the lower "idle" thresholds
+        vm["recommendationType"] = "Delete"
+        vm["recommendationDetails"] = [
+          "Delete Azure virtual machine ", vm["resourceId"], " ",
+          "in Azure Subscription ", vm["subscriptionName"], " ",
+          "(", vm["subscriptionId"], ")"
+        ].join("")
+        if (typeof vm["monthly_cost_estimate"] == "number") {
+          vm["savings"] = vm["monthly_cost_estimate"] // Potential Savings from delete is the estimated monthly cost of the resource
+        }
+        vm["memoryThreshold"] = param_stats_idle_threshold_mem_value
+        vm["cpuThreshold"] = param_stats_idle_threshold_cpu_value
+        idle_list.push(vm)
+      } else if (is_underutil) { // If not idle, check see if the resource is below the lower "idle" threshold
+        // Get the downsize instance type
+        if (ds_azure_instance_size_map[vm['resourceType']]) {
+          vm["newResourceType"] = ds_azure_instance_size_map[vm['resourceType']]['down']
+        }
+        vm["recommendationType"] = "Downsize"
+        vm["recommendationDetails"] = [
+          "Resize Azure virtual machine ", vm["resourceId"], " ",
+          "in Azure Subscription ", vm["subscriptionName"], " ",
+          "(", vm["subscriptionId"], ") ",
+          "from ", vm["resourceType"], " ",
+          "to ", vm["newResourceType"]
+        ].join("")
+        if (typeof vm["monthly_cost_estimate"] == "number") {
+          vm["savings"] = vm["monthly_cost_estimate"] / 2 // Potential Savings from downsize is generally half the estimated monthly cost of current resource type
+        }
+        vm["memoryThreshold"] = param_stats_underutil_threshold_mem_value
+        vm["cpuThreshold"] = param_stats_underutil_threshold_cpu_value
+        underutil_list.push(vm)
+      }
+    }
+  })
+  result = {
+    "underutil_list": underutil_list,
+    "idle_list": idle_list
+  }
+  EOS
+end
+
+datasource "ds_underutilized_instances" do
+  run_script $js_underutilized_instances, $ds_idle_and_underutil_instances
+end
+
+script "js_underutilized_instances", type: "javascript" do
+  parameters "ds_idle_and_underutil_instances"
+  result "result"
+  code <<-EOS
+  if (typeof ds_idle_and_underutil_instances["underutil_list"] != "undefined" && ds_idle_and_underutil_instances["underutil_list"].length > 0) {
```

Comment:
if ds_idle_and_underutil_instances["underutil_list"]  IS undefined, the second check will generate an error and cause the policy to fail because .length is not a valid method for objects of the type undefined

Can be fixed by just removing the length check entirely, since if the length is 0, "result = ds_idle_and_underutil_instances["underutil_list"]" and "result = []" will have the same result (heh) anyway

Reply:
https://github.com/flexera-public/policy_templates/blob/master/cost/azure/rightsize_compute_instances/azure_compute_rightsizing.pt#L1092-L1119
I recall initially using this approach from rightsize -- but I believe it didn't actually work, and I didn't bother troubleshooting because I already wanted to use another method that did not require statically defining the fields like we do in the result/export field section below.

JS will not attempt to process subsequent conditions in that if logic once it gets the `false` on the first statement -- 
```javascript
if (false && "string".toFixed(2)) {
	console.log("This should not run, or error.");
}
console.log("it worked!")
```
^^ plug this into [JSFiddle](https://jsfiddle.net/) to validate.  "string" objects don't have a `toFixed()` function and that would typically error if it was evaluated

I agree this is likely overkill but I think at this point in development I was constantly getting hung up on those small little exceptions and started to _check all the things.. everywhere_